yt/frontends/enzo_p/data_structures.py

```diff
@@ -0,0 +1,385 @@
+
+                bnl = nnl + 1
+                pbar.update(1)
+            lstr = buff[bnl:]
+            offset += fblock
+
+        f.close()
+        pbar.finish()
+
+        slope = self.ds.domain_width / \
+          self.ds.arr(np.ones(3), "code_length")
+        self.grid_left_edge   = self.grid_left_edge  * slope + \
+          self.ds.domain_left_edge
+        self.grid_right_edge  = self.grid_right_edge * slope + \
+          self.ds.domain_left_edge
+
+    def _populate_grid_objects(self):
+        for g in self.grids:
+            g._prepare_grid()
+            g._setup_dx()
+        self.max_level = self.grid_levels.max()
+
+    def _setup_derived_fields(self):
+        super(EnzoPHierarchy, self)._setup_derived_fields()
+        for fname, field in self.ds.field_info.items():
+            if not field.particle_type: continue
+            if isinstance(fname, tuple): continue
+            if field._function is NullFunc: continue
+
+    def _detect_output_fields(self):
+        self.field_list = []
+        # Do this only on the root processor to save disk work.
+        if self.comm.rank in (0, None):
+            mylog.info("Gathering a field list (this may take a moment.)")
+            field_list = set()
+            random_sample = self._generate_random_grids()
+            for grid in random_sample:
+                if not hasattr(grid, 'filename'): continue
+                try:
+                    gf = self.io._read_field_names(grid)
+                except self.io._read_exception:
+                    raise IOError("Grid %s is a bit funky?", grid.id)
+                mylog.debug("Grid %s has: %s", grid.id, gf)
+                field_list = field_list.union(gf)
+            if "AppendActiveParticleType" in self.dataset.parameters:
+                ap_fields = self._detect_active_particle_fields()
+                field_list = list(set(field_list).union(ap_fields))
+                if not any(f[0] == 'io' for f in field_list):
+                    if 'io' in self.dataset.particle_types_raw:
+                        ptypes_raw = list(self.dataset.particle_types_raw)
+                        ptypes_raw.remove('io')
+                        self.dataset.particle_types_raw = tuple(ptypes_raw)
+
+                    if 'io' in self.dataset.particle_types:
+                        ptypes = list(self.dataset.particle_types)
+                        ptypes.remove('io')
+                        self.dataset.particle_types = tuple(ptypes)
+            ptypes = self.dataset.particle_types
+            ptypes_raw = self.dataset.particle_types_raw
+        else:
+            field_list = None
+            ptypes = None
+            ptypes_raw = None
+        self.field_list = list(self.comm.mpi_bcast(field_list))
+        self.dataset.particle_types = list(self.comm.mpi_bcast(ptypes))
+        self.dataset.particle_types_raw = list(self.comm.mpi_bcast(ptypes_raw))
+
+    def _generate_random_grids(self):
+        if self.num_grids > 40:
```

Comment:
Enzo-P does not have a global list of fields, but this is something that I will push for when particle support is finalized.  That would remove the need for this, so I'll remove it.